### DISCUSSION TITLE
Cannot build GenericNonlinearExpr

### DISCUSSION THREAD

**Discussion by aml5600** [infiniteopt/InfiniteOpt.jl#354] [Q&A] [answered]
Hello! I am new to InfiniteOpt but very interested in solving an optimal control problem. I have an MTK model provided to me, from which I can generate functions to evaluate the system dynamics and constraint equations.

I have variables for the states, controls, and constraints:
```
@infinite_parameter(opt_model, tn in [0, 1], num_supports = 51, derivative_method=OrthogonalCollocation(4))
InfiniteOpt.@variable(opt_model, x_lb[i] <= x_opt[i = 1:nx] <= x_ub[i], Infinite(tn), start = guess_xs[i])
InfiniteOpt.@variable(opt_model, u_lb[i] <= u_opt[i = 1:nu] <= u_ub[i], Infinite(tn), start = guess_us[i])
InfiniteOpt.@variable(opt_model, c_lb[i] <= c_opt[i = 1:nc] <= c_ub[i], Infinite(tn))
```
and am trying to evaluate the dynamics function:
```
dx = f(x_opt, u_opt, p, tn)
```

but am seeing
```
ERROR: Cannot build `GenericNonlinearExpr` because a term is complex-valued: `(-1//2 * ((44.56410256410257 x_opt[8](tn) - 48.27777777777778 x_opt[7](tn) - 3645.513174722222) + -1.1025000009488324 * sqrt(x_opt[6](tn)² + x_opt[5](tn)²)^2) * 20)::NLPExpr`
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] _throw_if_not_real(x::NLPExpr)
   @ JuMP ~/.julia/packages/JuMP/7rBNn/src/nlp_expr.jl:334
 [3] *(x::NLPExpr, y::GenericNonlinearExpr{GeneralVariableRef})
   @ JuMP ~/.julia/packages/JuMP/7rBNn/src/nlp_expr.jl:391
 [4] macro expansion
   @ ~/.julia/packages/SymbolicUtils/EGhOJ/src/code.jl:468 [inlined]
 [5] macro expansion
   @ ~/.julia/packages/RuntimeGeneratedFunctions/M9ZX8/src/RuntimeGeneratedFunctions.jl:163 [inlined]
 [6] macro expansion
   @ ./none:0 [inlined]
 [7] generated_callfunc
   @ ./none:0 [inlined]
 [8] (::RuntimeGeneratedFunctions.RuntimeGeneratedFunction{…})(::Vector{…}, ::Vector{…}, ::Vector{…}, ::GeneralVariableRef)
   @ RuntimeGeneratedFunctions ~/.julia/packages/RuntimeGeneratedFunctions/M9ZX8/src/RuntimeGeneratedFunctions.jl:150
 [9] top-level scope
```

Not sure if relevant, but I have a) pointed some MTK functions back to Base: `ModelingToolkit.NaNMath.cos(x) = Base.cos(x)` and b) registered an interpolation function inside the MTK model.

Any tips on how to debug this? Thanks for the great package!

**Comment by pulsipher** [comment]
Hi there and welcome.

To better help you, would you please provide a minimum working example? From the code provided is not clear how `f` is produced and what you intend to do with `dx` if it were to work.

**Reply by aml5600** [reply]
The function `f` is generated by MTK from the model provided to me by:
```
(f, f_ip), dvs, psym, io_sys = ModelingToolkit.generate_control_function(sys, inputs, split = false);
```
where sys is an MTK `ODESystem`. This is the dynamics, so `x_dot = f(x, u, p, t)` where `x` is the state variable, `u` are inputs, and `p` are scalar parameters.

The plan is to enforce the dynamics as a constraint via:
```
@constraint(opt_model, [i = 1:nx], (∂((x_opt[i]), tn) - dx[i]) == 0)
```

I spent yesterday trying to recreate the issue with a MWE but was unable to isolate the issue. Unfortunately, I am unable to paste the full model (it is also spread across various files). I understand that is not helpful... =/

In general, are there typical failure cases where the `NLPExpr` is unable to be converted to Jump's `GenericNonlinearExpr`? Looking at the stacktrace, I cannot tell if the function is actually "not real", as the `_is_real` check is basically filtering out non-jump types?

**Reply by pulsipher** [reply, marked as answer]
Thanks for clarifying. The problem is likely that the current released version of InfiniteOpt has a legacy interface for nonlinear expressions based on `InfiniteOpt.NLPExpr` which may have issues with JuMP's new `GenericNonlinearExpr`s. The new version of InfiniteOpt, that will be released shortly, is updated to use `GenericNonlinearExpr`s.

To see if it fixes your problem, please try installing the current development version of InfiniteOpt:
```julia
using Pkg
Pkg.add(url = "https://github.com/infiniteopt/InfiniteOpt.jl", rev = "master")
```

**Reply by aml5600** [reply]
I'll give it a try, thank you!

**Comment by aml5600** [comment]
updating to master did the trick, thank you!!

**Reply by pulsipher** [reply]
Awesome, happy to help
